File: python-cli/src/convex_notes/repository/convex.py
"""Convex implementation of the NoteRepository."""

from __future__ import annotations

from typing import TYPE_CHECKING, Iterator

from convex import ConvexClient

from convex_notes.entity import Note

if TYPE_CHECKING:
    from convex_notes.entity import CreateNote, UpdateNote


class ConvexNoteRepository:
    """Convex implementation of the NoteRepository protocol."""

    def __init__(self, client: ConvexClient) -> None:
        """Initialize with a Convex client."""
        self._client = client

    def list(self) -> list[Note]:
        """List all notes from Convex."""
        result = self._client.query("notes:list", {})
        return [Note.from_convex(item) for item in result]

    def get(self, note_id: str) -> Note | None:
        """Get a note by ID from Convex."""
        result = self._client.query("notes:get", {"id": note_id})
        if result is None:
            return None
        return Note.from_convex(result)

    def create(self, note: CreateNote) -> str:
        """Create a new note in Convex."""
        result = self._client.mutation(
            "notes:create",
            {"title": note.title, "content": note.content},
        )
        return str(result)

    def update(self, note: UpdateNote) -> None:
        """Update an existing note in Convex."""
        self._client.mutation(
            "notes:update",
            {"id": note.id, "title": note.title, "content": note.content},
        )

    def delete(self, note_id: str) -> None:
        """Delete a note from Convex."""
        self._client.mutation("notes:remove", {"id": note_id})

    def subscribe(self) -> Iterator[list[Note]]:
        """Subscribe to note changes from Convex."""
        for result in self._client.subscribe("notes:list", {}):
            yield [Note.from_convex(item) for item in result]

File: bun-cli/src/repository/convex.ts
/**
 * Convex implementation of NoteRepository using Effect.Service.
 */

import { ConvexClient, ConvexHttpClient } from "convex/browser";
import { Effect } from "effect";
import { AppConfig } from "../config";
import type { CreateNote, Note, NoteId, UpdateNote } from "../entity/note";
import { parseNote } from "../entity/note";
import { ConvexError } from "../errors";
import type { NoteRepositoryShape } from "./interface";

/**
 * NoteRepository service using Effect.Service pattern.
 */
export class NoteRepository extends Effect.Service<NoteRepository>()(
  "NoteRepository",
  {
    effect: Effect.gen(function* () {
      const config = yield* AppConfig;
      const httpClient = new ConvexHttpClient(config.convexUrl);

      // Lazy WebSocket client
      let wsClient: ConvexClient | null = null;
      const getWsClient = () => {
        if (!wsClient) {
          wsClient = new ConvexClient(config.convexUrl);
        }
        return wsClient;
      };

      const repo: NoteRepositoryShape = {
        list: () =>
          Effect.tryPromise({
            try: async () => {
              const result = await httpClient.query("notes:list" as any, {});
              return (result as unknown[]).map((item) =>
                parseNote(item as Record<string, unknown>),
              );
            },
            catch: (error) =>
              new ConvexError({
                message: String(error),
                operation: "list",
                cause: error,
              }),
          }),

        get: (id: NoteId) =>
          Effect.tryPromise({
            try: async () => {
              const result = await httpClient.query("notes:get" as any, { id });
              if (!result) return null;
              return parseNote(result as Record<string, unknown>);
            },
            catch: (error) =>
              new ConvexError({
                message: String(error),
                operation: "get",
                cause: error,
              }),
          }),

        create: (note: CreateNote) =>
          Effect.tryPromise({
            try: async () => {
              const result = await httpClient.mutation("notes:create" as any, {
                title: note.title,
                content: note.content,
              });
              return String(result) as NoteId;
            },
            catch: (error) =>
              new ConvexError({
                message: String(error),
                operation: "create",
                cause: error,
              }),
          }),

        update: (note: UpdateNote) =>
          Effect.tryPromise({
            try: async () => {
              await httpClient.mutation("notes:update" as any, {
                id: note.id,
                title: note.title,
                content: note.content,
              });
            },
            catch: (error) =>
              new ConvexError({
                message: String(error),
                operation: "update",
                cause: error,
              }),
          }),

        delete: (id: NoteId) =>
          Effect.tryPromise({
            try: async () => {
              await httpClient.mutation("notes:remove" as any, { id });
            },
            catch: (error) =>
              new ConvexError({
                message: String(error),
                operation: "delete",
                cause: error,
              }),
          }),

        subscribe: (callback: (notes: Note[]) => void) =>
          Effect.sync(() => {
            return getWsClient().onUpdate("notes:list" as any, {}, (result) => {
              const notes = (result as unknown[]).map((item) =>
                parseNote(item as Record<string, unknown>),
              );
              callback(notes);
            });
          }),
      };

      return repo;
    }),
    dependencies: [AppConfig.Default],
  },
) {}
